quinn-proto/src/transport_parameters.rs

```diff
@@ -94,6 +94,67 @@ macro_rules! make_struct {
             pub(crate) preferred_address: Option<PreferredAddress>,
         }
 
+        impl TransportParameters {
+            $($(#[$doc])*
+            #[inline]
+            pub fn $name (&self) -> u64 {
+                self.$name.0
+            })*
+
+            /// Does the endpoint support active connection migration
+            #[inline]
+            pub fn disable_active_migration(&self) -> bool {
+                self.disable_active_migration
+            }
+
+            /// Maximum size for datagram frames
+            #[inline]
+            pub fn max_datagram_frame_size(&self) -> Option<u64> {
+                self.max_datagram_frame_size.map_or(None, |x| Some(x.0))
+            }
+
+            /// The value that the endpoint included in the Source Connection ID field of the first
+            /// Initial packet it sends for the connection
+            #[inline]
+            pub fn initial_src_cid(&self) -> Option<ConnectionId> {
+                self.initial_src_cid
+            }
+
+            /// The endpoint is willing to receive QUIC packets containing any value for the fixed
+            /// bit
+            #[inline]
+            pub fn grease_quic_bit(&self) -> bool {
+                self.grease_quic_bit
+            }
+
+            // Server-only
+            /// The value of the Destination Connection ID field from the first Initial packet sent
+            /// by the client
+            #[inline]
+            pub fn original_dst_cid(&self) -> Option<ConnectionId> {
+                self.original_dst_cid
+            }
+
+            /// The value that the server included in the Source Connection ID field of a Retry
+            /// packet
+            #[inline]
+            pub fn retry_src_cid(&self) -> Option<ConnectionId> {
+                self.retry_src_cid
+            }
+
+            /// Token used by the client to verify a stateless reset from the server
+            #[inline]
+            pub fn stateless_reset_token(&self) -> Option<ResetToken> {
+                self.stateless_reset_token
+            }
+
+            /// The server's preferred address for communication after handshake completion
+            #[inline]
+            pub fn preferred_address(&self) -> Option<PreferredAddress> {
+                self.preferred_address
+            }
+        }
+
```

Comment:
Can you elaborate on the motivation for exposing these? Most of these are used purely by the transport layer, IIRC.

Reply:
Wouldn't it be better to use the `TransportParameters::read()` and `write()` methods, which are already public? Looks like that API needs a byte buffer.

Reply:
Only select fields of transport parameters are included in the early data context, so I don't read/write will work.

Reply:
Have you found code that does the equivalent of this on the rustls code path? All other things being equal, adding code for this to `TransportParameters` seems like it could make sense.

Reply:
I copied the logic from Google's quiche library: https://github.com/google/quiche/blob/6a93efcba4e4339bb3b4fe14f340f4e3f60b28c7/quiche/quic/core/crypto/transport_parameters.cc#L1553

Reply:
@Ralith huh, do you remember how/where this is implemented for our rustls code path?

Reply:
Interesting; BoringSSL is taking responsibility for not accepting 0-RTT on the server if the server's configuration has changed, as a conservative solution. Our current behavior is to not check this; it's the user's responsibility not to change transport parameters in illegal ways without also invalidating resumption tokens via rustls. Quiche's approach is a bit less error prone, but also more restrictive than is strictly necessary since it prevents relaxing flow control limits.

A quick and easy way to make BoringSSL happy here would be to pass an empty or fixed context. I don't see a security risk to doing so as far as QUIC is concerned; the biggest hazard is that someone will illegally reduce a limit and cause connections resumed across the change to fail with `PROTOCOL_VIOLATION`. Perhaps there's a larger concern for higher-level protocols, but those can customize the rustls behavior directly already.

We could match Quiche's behavior without as much API expansion by offering an accessor for an opaque byte string, but I'm not convinced even that much is necessary. @nmittler do you have an opinion?